File: crates/czdev/src/publish.rs
use anyhow::{anyhow, Context, Result};
use rust_i18n::t;
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::auth;
use crate::github::{GitHubClient, Permission};

const TARGET_OWNER: &str = "CardputerZero";
const TARGET_REPO: &str = "packages";

struct DebMetadata {
    package: String,
    version: String,
    architecture: String,
    maintainer: String,
    maintainer_email: String,
}

pub fn run(deb: Option<&Path>) -> Result<()> {
    let deb_path = resolve_deb(deb)?;
    println!("Package: {}", deb_path.display());
    println!();

    let token = auth::load_token()?;
    let gh = GitHubClient::new(&token);

    let user = gh.get_user().context("fetching user info")?;

    // Build list of emails to match against: public email + noreply
    // (no need for user:email scope)
    let noreply = format!("{}@users.noreply.github.com", user.login);
    let mut all_emails = vec![noreply];
    if let Some(ref email) = user.email {
        if !email.is_empty() {
            all_emails.push(email.clone());
        }
    }

    println!("{}", t!("publish.preflight"));

    // 1. Check .desktop file exists
    let has_desktop = check_desktop(&deb_path)?;
    if !has_desktop {
        return Err(anyhow!("{}", t!("publish.desktop_missing")));
    }
    println!("  ✓ {}", t!("publish.desktop_found"));

    // 2. Extract metadata and check email
    let meta = extract_metadata(&deb_path)?;
    if !all_emails.iter().any(|e| e.eq_ignore_ascii_case(&meta.maintainer_email)) {
        return Err(anyhow!(
            "{}\n  Maintainer: {}\n  Your emails: {:?}",
            t!("publish.email_mismatch"),
            meta.maintainer_email,
            all_emails
        ));
    }
    println!("  ✓ {}", t!("publish.email_match"));

    // 3. Package name validation
    if !is_valid_package_name(&meta.package) {
        return Err(anyhow!("{}: '{}'", t!("publish.pkg_invalid"), meta.package));
    }
    println!("  ✓ {} \"{}\"", t!("publish.pkg_valid"), meta.package);

    // 4. Show summary
    let file_size = std::fs::metadata(&deb_path)?.len();
    let size_mb = file_size as f64 / 1_048_576.0;
    println!(
        "  ✓ Version: {}, Arch: {}, Size: {:.1} MB",
        meta.version, meta.architecture, size_mb
    );
    println!();

    if file_size > 100 * 1024 * 1024 {
        return Err(anyhow!("{} ({:.1} MB)", t!("publish.too_large"), size_mb));
    }

    // 5. Check version is newer than existing
    check_version_newer(&gh, &meta)?;

    // Determine target: direct push or fork
    let perm = gh.check_permission(TARGET_OWNER, TARGET_REPO, &user.login)?;
    let (push_owner, push_repo, pr_head) = if perm >= Permission::Write {
        (TARGET_OWNER.to_string(), TARGET_REPO.to_string(), None)
    } else {
        println!("{} {TARGET_OWNER}/{TARGET_REPO}.", t!("publish.no_write_access"));
        print!("  → {}  ", t!("publish.forking"));
        let fork_name = gh.fork_repo(TARGET_OWNER, TARGET_REPO)?;
        println!("done ({fork_name})");
        let parts: Vec<&str> = fork_name.split('/').collect();
        (
            parts[0].to_string(),
            parts[1].to_string(),
            Some(format!("{}:{}", user.login, branch_name(&meta))),
        )
    };

    println!("{} {TARGET_OWNER}/{TARGET_REPO}...", t!("publish.uploading_to"));

    // Get base ref
    let base_sha = gh.get_ref_sha(&push_owner, &push_repo, "heads/main")?;
    let (_, base_tree_sha) = gh.get_commit(&push_owner, &push_repo, &base_sha)?;

    // Upload via LFS + create pointer blob
    print!("  → {} ({:.1} MB)... ", t!("publish.uploading_blob"), size_mb);
    let file_bytes = std::fs::read(&deb_path).context("reading deb file")?;
    let sha256_hash = hex_sha256(&file_bytes);
    let blob_sha = gh.upload_lfs_and_create_pointer_blob(
        &push_owner, &push_repo, &file_bytes, &sha256_hash,
    )?;
    println!("{} (sha: {})", t!("publish.done"), &blob_sha[..8]);

    // Create tree
    let file_path_in_repo = format!(
        "pool/main/{}/{}_{}_{}.deb",
        meta.package, meta.package, meta.version, meta.architecture
    );
    print!("  → {} ", t!("publish.creating_tree"));
    let tree_sha =
        gh.create_tree(&push_owner, &push_repo, &base_tree_sha, &file_path_in_repo, Some(&blob_sha))?;
    println!("{}", t!("publish.done"));

    // Create commit
    let commit_msg = format!("publish: {} {} ({})", meta.package, meta.version, meta.architecture);
    print!("  → {} ", t!("publish.creating_commit"));
    let commit_sha = gh.create_commit(&push_owner, &push_repo, &commit_msg, &tree_sha, &base_sha)?;
    println!("{}", t!("publish.done"));

    // Create branch
    let branch = branch_name(&meta);
    print!("  → {} {branch}... ", t!("publish.creating_branch"));
    gh.create_ref(&push_owner, &push_repo, &branch, &commit_sha)?;
    println!("{}", t!("publish.done"));

    // Create PR
    let head = pr_head.unwrap_or_else(|| branch.clone());
    let pr_body = format!(
        "## Package: `{}`\n\n\
         | Field | Value |\n\
         |-------|-------|\n\
         | Version | {} |\n\
         | Architecture | {} |\n\
         | Maintainer | {} |\n\
         | Size | {:.1} MB |\n\
         | SHA-256 | `{}` |\n\
         | File | `{}` |\n\n\
         Submitted via `czdev publish`.",
        meta.package,
        meta.version,
        meta.architecture,
        meta.maintainer,
        size_mb,
        sha256_hash,
        file_path_in_repo,
    );
    print!("  → {} ", t!("publish.creating_pr"));
    let pr = gh.create_pull_request(
        TARGET_OWNER,
        TARGET_REPO,
        &format!("publish: {} {}", meta.package, meta.version),
        &pr_body,
        &head,
        "main",
    )?;
    println!("{}", t!("publish.done"));

    println!();
    println!("✓ {}", t!("publish.pr_created"));
    println!("  {}", pr.html_url);
    println!();
    println!("  {}", t!("publish.pr_hint"));
    Ok(())
}

fn resolve_deb(deb: Option<&Path>) -> Result<PathBuf> {
    if let Some(p) = deb {
        if !p.is_file() {
            return Err(anyhow!("{} {}", t!("common.file_not_found"), p.display()));
        }
        return Ok(p.to_path_buf());
    }
    // Search build/ for a .deb
    let build_dir = Path::new("build");
    if build_dir.is_dir() {
        let mut debs: Vec<_> = std::fs::read_dir(build_dir)?
            .filter_map(|e| e.ok())
            .filter(|e| e.path().extension().map(|x| x == "deb").unwrap_or(false))
            .collect();
        if debs.len() == 1 {
            return Ok(debs.remove(0).path());
        }
        if debs.len() > 1 {
            return Err(anyhow!(
                "multiple .deb files in build/. Specify one with --deb <path>"
            ));
        }
    }
    Err(anyhow!("no .deb file found. Specify with --deb <path>"))
}

fn check_desktop(deb: &Path) -> Result<bool> {
    let output = Command::new("dpkg-deb")
        .arg("-c")
        .arg(deb)
        .output()
        .context("running dpkg-deb -c (is dpkg-deb installed?)")?;
    let listing = String::from_utf8_lossy(&output.stdout);
    Ok(listing.lines().any(|l| l.ends_with(".desktop")))
}

fn extract_metadata(deb: &Path) -> Result<DebMetadata> {
    let fields = &["Package", "Version", "Architecture", "Maintainer"];
    let mut values = std::collections::HashMap::new();

    for field in fields {
        let output = Command::new("dpkg-deb")
            .args(["-f", &deb.to_string_lossy(), field])
            .output()
            .with_context(|| format!("dpkg-deb -f {field}"))?;
        let val = String::from_utf8_lossy(&output.stdout).trim().to_string();
        values.insert(*field, val);
    }

    let maintainer = values.get("Maintainer").cloned().unwrap_or_default();
    let email = extract_email(&maintainer);

    Ok(DebMetadata {
        package: values.get("Package").cloned().unwrap_or_default(),
        version: values.get("Version").cloned().unwrap_or_default(),
        architecture: values.get("Architecture").cloned().unwrap_or_default(),
        maintainer,
        maintainer_email: email,
    })
}

fn extract_email(maintainer: &str) -> String {
    if let Some(start) = maintainer.find('<') {
        if let Some(end) = maintainer.find('>') {
            return maintainer[start + 1..end].to_string();
        }
    }
    maintainer.to_string()
}

fn is_valid_package_name(name: &str) -> bool {
    if name.len() < 2 {
        return false;
    }
    let bytes = name.as_bytes();
    (bytes[0].is_ascii_lowercase() || bytes[0].is_ascii_digit())
        && bytes.iter().all(|&b| {
            b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'.' || b == b'+' || b == b'-'
        })
}

fn branch_name(meta: &DebMetadata) -> String {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs();
    format!("publish/{}-{}-{}", meta.package, meta.version, ts)
}

fn hex_sha256(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    format!("{:x}", hasher.finalize())
}

fn check_version_newer(_gh: &GitHubClient, meta: &DebMetadata) -> Result<()> {
    // Try to fetch the Packages index to see if this package already exists
    let packages_url = format!(
        "https://cardputerzero.github.io/packages/dists/stable/main/binary-arm64/Packages"
    );
    let resp = reqwest::blocking::get(&packages_url);
    let content = match resp {
        Ok(r) if r.status().is_success() => r.text().unwrap_or_default(),
        _ => return Ok(()), // Can't check, skip (repo might be empty)
    };

    // Parse existing version for this package
    let mut in_our_package = false;
    let mut existing_version: Option<String> = None;
    for line in content.lines() {
        if line.starts_with("Package: ") {
            in_our_package = line.trim_start_matches("Package: ") == meta.package;
        }
        if in_our_package && line.starts_with("Version: ") {
            let ver = line.trim_start_matches("Version: ").to_string();
            // Keep the highest version found
            match &existing_version {
                Some(ev) if compare_versions(&ver, ev) == std::cmp::Ordering::Greater => {
                    existing_version = Some(ver);
                }
                None => existing_version = Some(ver),
                _ => {}
            }
        }
        if line.is_empty() {
            in_our_package = false;
        }
    }

    if let Some(existing) = existing_version {
        if compare_versions(&meta.version, &existing) != std::cmp::Ordering::Greater {
            return Err(anyhow!(
                "{} {} {} {}",
                meta.version, t!("publish.version_not_newer"), existing,
                "\n  Use `czdev bump` to check the next version."
            ));
        }
        println!("  ✓ {} {} {} {}", meta.version, t!("publish.version_newer"), existing, "");
    } else {
        println!("  ✓ {}", t!("publish.new_package"));
    }
    Ok(())
}

fn compare_versions(a: &str, b: &str) -> std::cmp::Ordering {
    // Simple version comparison: split by '.', '-', '~' and compare segments
    let parse = |v: &str| -> Vec<u64> {
        v.split(|c: char| c == '.' || c == '-' || c == '~')
            .filter_map(|s| s.parse::<u64>().ok())
            .collect()
    };
    let va = parse(a);
    let vb = parse(b);
    va.cmp(&vb)
}

File: scripts/czdev/publish.py
"""Publish a .deb package to the CardputerZero app store — mirrors the Rust publish module."""

import hashlib
import json
import os
import re
import shutil
import subprocess
import sys
import tempfile
import time
import urllib.request
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from . import auth
from .github_client import GitHubClient, Permission

TARGET_OWNER = "CardputerZero"
TARGET_REPO = "packages"
# Buffer release tag on the push target (contributor fork, or the official repo
# for maintainers) where the .deb is uploaded before review.
BUFFER_RELEASE_TAG = "czdev-buffer"


def run(deb: Optional[str] = None):
    check_git_installed()

    deb_path = resolve_deb(deb)
    print(f"Package: {deb_path}")

    store_meta, app_dir = load_store_meta(deb_path)
    print()

    token = auth.load_token()
    gh = GitHubClient(token)

    user = gh.get_user()

    noreply = f"{user.login}@users.noreply.github.com"
    uploader_email = user.email or noreply

    print("Preflight checks:")

    # 1. Check .desktop file exists
    if not check_desktop(deb_path):
        print("ERROR: deb does not contain a .desktop file. All CardputerZero apps must include one.", file=sys.stderr)
        sys.exit(1)
    print("  ✓ .desktop file found")

    # 1b. Apps must not run as root. Reject any bundled systemd *system* service
    #     that would run as root (no User=, User=root/0, and no DynamicUser=yes).
    root_services = find_root_services(deb_path)
    if root_services:
        print("ERROR: apps must not run as root, but these systemd services run as root:", file=sys.stderr)
        for s in root_services:
            print(f"         - {s}", file=sys.stderr)
        print("       Add `User=<non-root>` (e.g. `User=pi`) to the [Service] section, or", file=sys.stderr)
        print("       rebuild with `pack_deb.py --service-user pi` / `--no-service`.", file=sys.stderr)
        sys.exit(1)
    print("  ✓ no root systemd services")

    # 2. Extract metadata. Ownership is first-come-first-served by package name,
    #    keyed on the uploader's GitHub login (recorded in the release manifest
    #    below and enforced server-side). We no longer require the deb's
    #    Maintainer email to match the uploader's GitHub emails — we just record
    #    who uploaded it.
    meta = extract_metadata(deb_path)
    print(f"  ✓ Maintainer: {meta['maintainer'] or meta['maintainer_email'] or '(none)'}")

    # 3. Package name validation
    if not is_valid_package_name(meta["package"]):
        print(f"ERROR: Invalid package name: '{meta['package']}'", file=sys.stderr)
        sys.exit(1)
    print(f'  ✓ Package name is valid "{meta["package"]}"')

    # 4. Show summary
    file_size = os.path.getsize(deb_path)
    size_mb = file_size / 1_048_576.0
    print(f"  ✓ Version: {meta['version']}, Arch: {meta['architecture']}, Size: {size_mb:.1f} MB")
    print()

    if file_size > 2 * 1024 * 1024 * 1024:
        print(f"ERROR: File too large. GitHub release asset limit is 2 GiB. ({size_mb:.1f} MB)", file=sys.stderr)
        sys.exit(1)

    # 5. Check version is newer than existing
    check_version_newer(meta)

    # 6. Build the meta.json that will be committed and pre-check it against
    #    the store's own policy script BEFORE uploading anything, so a
    #    submission the CI would reject fails right here with the same wording.
    meta_out = build_merged_meta(store_meta, meta, user)
    precheck_store_meta(meta_out, app_dir, store_meta, meta["package"])

    # Determine target: direct push or fork
    perm = gh.check_permission(TARGET_OWNER, TARGET_REPO)
    if perm >= Permission.WRITE:
        push_owner = TARGET_OWNER
        push_repo = TARGET_REPO
        pr_head = None
    else:
        print(f"You don't have write access to {TARGET_OWNER}/{TARGET_REPO}.")
        print("  → Forking to your account...  ", end="", flush=True)
        fork_name = gh.fork_repo(TARGET_OWNER, TARGET_REPO)
        print(f"done ({fork_name})")
        parts = fork_name.split("/")
        push_owner = parts[0]
        push_repo = parts[1]
        branch = branch_name(meta)
        pr_head = f"{user.login}:{branch}"

    # Integrity + canonical asset name. GitHub sanitizes '~' to '.' in release
    # asset names, so the on-release/manifest filename may differ from the
    # Debian version string (which keeps '~').
    file_bytes = Path(deb_path).read_bytes()
    sha256_hash = hashlib.sha256(file_bytes).hexdigest()
    file_size = len(file_bytes)
    deb_name = f"{meta['package']}_{meta['version']}_{meta['architecture']}.deb"
    asset_name = deb_name.replace("~", ".")
    manifest_name = f"{asset_name}.release.json"
    manifest_path_in_repo = f"pool/main/{meta['package']}/{manifest_name}"
    branch = branch_name(meta)
    remote_url = f"git@github.com:{push_owner}/{push_repo}.git"

    # 1) Upload the .deb to a buffer Release on the push target. For third-party
    #    contributors this is their own fork — it uses THEIR free Release storage
    #    and bandwidth, so the upstream project stores nothing until the PR is
    #    approved and CI promotes the binary into the official apt-pool release.
    print(f"  → Uploading .deb ({size_mb:.1f} MB) to {push_owner}/{push_repo} release '{BUFFER_RELEASE_TAG}'... ",
          end="", flush=True)
    release = gh.ensure_release(push_owner, push_repo, BUFFER_RELEASE_TAG, name="czdev upload buffer")
    download_url = gh.upload_release_asset(push_owner, push_repo, release, deb_path, asset_name)
    print("done")

    manifest = {
        "filename": asset_name,
        "url": download_url,
        "sha256": sha256_hash,
        "size": file_size,
        "package": meta["package"],
        "version": meta["version"],
        "architecture": meta["architecture"],
        # Ownership record: first uploader of a package name owns it. The login
        # is the authoritative owner key; the email is kept for contact/audit.
        "uploaded_by": user.login,
        "uploader_email": uploader_email,
    }

    # 2) Commit only metadata (meta.json, screenshots, icon, manifest) to a PR
    #    branch — no .deb, no LFS.
    print(f"Uploading metadata to {TARGET_OWNER}/{TARGET_REPO}...")

    tmp_dir = Path(tempfile.mkdtemp(prefix="czdev-publish-"))
    try:
        run_cmd_in(tmp_dir, ["git", "init"])
        run_cmd_in(tmp_dir, ["git", "remote", "add", "origin", remote_url])

        print("  → git fetch (minimal)... ", end="", flush=True)
        run_cmd_in(tmp_dir, ["git", "fetch", "--depth=1", "--filter=blob:none", "origin", "main"])
        print("done")

        print(f"  → Creating branch {branch}... ", end="", flush=True)
        run_cmd_in(tmp_dir, ["git", "checkout", "-b", branch, "origin/main"])
        print("done")

        dest_dir = tmp_dir / "pool" / "main" / meta["package"]
        dest_dir.mkdir(parents=True, exist_ok=True)

        # Copy screenshots
        if store_meta.get("screenshots"):
            screenshots_dest = dest_dir / "screenshots"
            screenshots_dest.mkdir(parents=True, exist_ok=True)
            for shot in store_meta["screenshots"]:
                src = Path(app_dir) / shot
                if not src.exists():
                    print(f"ERROR: Screenshot not found: {src}", file=sys.stderr)
                    sys.exit(1)
                shutil.copy2(src, screenshots_dest / src.name)

        # Copy icon
        if store_meta.get("icon"):
            icon_src = Path(app_dir) / store_meta["icon"]
            if icon_src.exists():
                shutil.copy2(icon_src, dest_dir / icon_src.name)

        # meta.json + release manifest (the binary stays in the buffer release).
        # meta_out was built and pre-checked before the upload.
        (dest_dir / "meta.json").write_text(json.dumps(meta_out, indent=2, ensure_ascii=False) + "\n")
        (dest_dir / manifest_name).write_text(json.dumps(manifest, indent=2) + "\n")

        print("  → Creating commit... ", end="", flush=True)
        run_cmd_in(tmp_dir, ["git", "add", f"pool/main/{meta['package']}"])
        run_cmd_in(tmp_dir, ["git", "commit", "-m",
                             f"publish: {meta['package']} {meta['version']} ({meta['architecture']})"])
        print("done")

        print("  → Pushing branch... ", end="", flush=True)
        run_cmd_in(tmp_dir, ["git", "push", "origin", branch])
        print("done")

    finally:
        shutil.rmtree(tmp_dir, ignore_errors=True)

    # Create PR via API
    head = pr_head if pr_head else branch
    pr_body = (
        f"## Package: `{meta['package']}`\n\n"
        f"| Field | Value |\n"
        f"|-------|-------|\n"
        f"| Version | {meta['version']} |\n"
        f"| Architecture | {meta['architecture']} |\n"
        f"| Maintainer | {meta['maintainer']} |\n"
        f"| Size | {size_mb:.1f} MB |\n"
        f"| SHA-256 | `{sha256_hash}` |\n"
        f"| Manifest | `{manifest_path_in_repo}` |\n"
        f"| Binary | [{asset_name}]({download_url}) |\n\n"
        f"Submitted via `czdev publish`. The `.deb` is hosted on the contributor's "
        f"buffer release; CI verifies the sha256 and, on merge, promotes it into the "
        f"official `apt-pool` release."
    )
    print("  → Creating pull request... ", end="", flush=True)
    pr = gh.create_pull_request(
        TARGET_OWNER, TARGET_REPO,
        f"publish: {meta['package']} {meta['version']}",
        pr_body, head, "main",
    )
    print("done")

    print()
    print("✓ Pull request created:")
    print(f"  {pr.html_url}")
    print()
    print("  The PR will be validated by CI. A maintainer will review and merge it.")


def resolve_deb(deb: Optional[str]) -> str:
    if deb:
        if not os.path.isfile(deb):
            print(f"File not found: {deb}", file=sys.stderr)
            sys.exit(1)
        return deb
    build_dir = Path("build")
    if build_dir.is_dir():
        debs = list(build_dir.glob("*.deb"))
        if len(debs) == 1:
            return str(debs[0])
        if len(debs) > 1:
            print("multiple .deb files in build/. Specify one with --deb <path>", file=sys.stderr)
            sys.exit(1)
    print("no .deb file found. Specify with --deb <path>", file=sys.stderr)
    sys.exit(1)


def load_store_meta(deb_path: str) -> tuple:
    search_dirs = [
        Path.cwd(),
        Path(deb_path).parent,
    ]

    for d in search_dirs:
        manifest_path = d / "app-builder.json"
        if manifest_path.exists():
            raw = json.loads(manifest_path.read_text())

            if "store" not in raw:
                print("app-builder.json found but missing \"store\" section.", file=sys.stderr)
                print('  Add a "store" field with title, summary, categories, screenshots, etc.', file=sys.stderr)
                sys.exit(1)

            store = raw["store"]
            meta = {
                "title": raw.get("app_name", ""),
                "summary": store.get("summary", ""),
                "categories": store.get("categories", []),
                "screenshots": store.get("screenshots", []),
            }
            for key in ("description", "locales", "license", "source_repo",
                        "icon", "permissions", "share_code"):
                if store.get(key):
                    meta[key] = store[key]
            if isinstance(store.get("author"), dict):
                meta["author"] = store["author"]

            if not meta["screenshots"]:
                print("No screenshots defined in app-builder.json store.screenshots.", file=sys.stderr)
                print("  Add at least one 320x170 screenshot to publish.", file=sys.stderr)
                sys.exit(1)

            return meta, str(d)

    print("app-builder.json not found in current directory.", file=sys.stderr)
    print("  Run `czdev publish` from your app's project directory.", file=sys.stderr)
    sys.exit(1)


RAW_BASE = f"https://raw.githubusercontent.com/{TARGET_OWNER}/{TARGET_REPO}/main"


def fetch_url(url: str, timeout: int = 15) -> Optional[bytes]:
    try:
        return urllib.request.urlopen(urllib.request.Request(url), timeout=timeout).read()
    except Exception:
        return None


def build_merged_meta(store_meta: dict, deb_meta: dict, user) -> dict:
    """The meta.json to publish: the already-published metadata as the base,
    the app-builder.json store section on top.

    Merging matters: fields the store section does not carry — most notably
    the uuid the registry pinned, or a share_code assigned in an earlier
    backfill — must survive an update instead of being wiped by a rewrite.
    Asset paths are rewritten to the pool layout the files are copied into
    (icon at <pkg-dir>/<basename>, screenshots under <pkg-dir>/screenshots/) —
    the app-builder.json values are source-repo relative and would 404 when
    the store resolves them against pool/main/<pkg>/.
    """
    merged = {}
    published = fetch_url(f"{RAW_BASE}/pool/main/{deb_meta['package']}/meta.json")
    if published:
        try:
            merged = json.loads(published.decode("utf-8"))
        except Exception:
            merged = {}

    update = dict(store_meta)
    if update.get("icon"):
        update["icon"] = Path(update["icon"]).name
    if update.get("screenshots"):
        update["screenshots"] = [f"screenshots/{Path(s).name}" for s in update["screenshots"]]
    author = update.pop("author", None)
    for k, v in update.items():
        if v not in (None, "", [], {}):
            merged[k] = v
    merged_author = dict(merged.get("author") or {})
    if isinstance(author, dict):
        merged_author.update({k: v for k, v in author.items() if v})
    merged_author["github"] = user.login
    if not merged_author.get("display_name"):
        display_name = deb_meta["maintainer"].split("<")[0].strip()
        merged_author["display_name"] = display_name or user.login
    merged["author"] = merged_author
    # Publish timestamps: the registry renders these; without them it falls
    # back to overrides or the registry build time, both wrong for updates.
    now = datetime.now(timezone.utc).isoformat(timespec="seconds")
    merged.setdefault("published_at", now)
    merged["updated_at"] = now
    return merged


def precheck_store_meta(meta_out: dict, app_dir: str, store_meta: dict, pkg: str):
    """Run the store's own metadata policy locally before uploading anything.

    Uses the exact script CI runs (fetched from the packages repo), staged
    against a temporary pool/main/<pkg>/ layout, so the verdict and the wording
    match what the PR would get. Uniqueness needs the whole pool and is left
    to CI. If the script cannot be fetched (offline), the pre-check is skipped
    and CI remains the authority.
    """
    script = fetch_url(f"{RAW_BASE}/.github/scripts/store_meta_policy.py")
    if not script:
        print("  ⚠ could not fetch store_meta_policy.py — skipping local pre-check (CI will validate)")
        return

    stage = Path(tempfile.mkdtemp(prefix="czdev-precheck-"))
    try:
        pkg_dir = stage / pkg
        (pkg_dir / "screenshots").mkdir(parents=True)
        (pkg_dir / "meta.json").write_text(json.dumps(meta_out, ensure_ascii=False))
        for shot in store_meta.get("screenshots") or []:
            src = Path(app_dir) / shot
            if src.is_file():
                shutil.copy2(src, pkg_dir / "screenshots" / src.name)
        if store_meta.get("icon"):
            icon_src = Path(app_dir) / store_meta["icon"]
            if icon_src.is_file():
                shutil.copy2(icon_src, pkg_dir / icon_src.name)

        script_path = stage / "store_meta_policy.py"
        script_path.write_bytes(script)
        errors_path = stage / "errors.txt"
        warnings_path = stage / "warnings.txt"
        subprocess.run([sys.executable, str(script_path), pkg, str(pkg_dir), "-",
                        str(errors_path), str(warnings_path)], check=False)
        if warnings_path.is_file():
            for line in warnings_path.read_text(encoding="utf-8").splitlines():
                print(f"  ⚠ {line}")
        if errors_path.is_file():
            print("ERROR: store metadata does not meet the store policy:", file=sys.stderr)
            for line in errors_path.read_text(encoding="utf-8").splitlines():
                print(f"  ✗ {line}", file=sys.stderr)
            print("\n  Fix the \"store\" section of app-builder.json (and the referenced", file=sys.stderr)
            print("  images), then run `czdev publish` again. Nothing was uploaded.", file=sys.stderr)
            sys.exit(1)
        print("  ✓ store metadata passes the store policy")
    finally:
        shutil.rmtree(stage, ignore_errors=True)


# Directories where a *system* systemd unit runs as root by default (user
# units under .../systemd/user/ run as the logged-in user and are not a concern).
_SYSTEMD_SYSTEM_DIRS = ("lib/systemd/system/", "usr/lib/systemd/system/", "etc/systemd/system/")


def service_runs_as_root(text: str) -> bool:
    """Return True if a systemd .service unit would run as root.

    A unit runs as root unless its [Service] section pins it to a non-root
    identity via `User=<non-root>` or `DynamicUser=yes`. An empty/`root`/`0`
    User= (or none) means root.
    """
    in_service = False
    user = None
    dynamic = False
    for raw in text.splitlines():
        line = raw.strip()
        if not line or line.startswith(("#", ";")):
            continue
        if line.startswith("[") and line.endswith("]"):
            in_service = line.lower() == "[service]"
            continue
        if not in_service or "=" not in line:
            continue
        key, _, val = line.partition("=")
        key = key.strip().lower()
        val = val.strip()
        if key == "user":
            user = val
        elif key == "dynamicuser":
            dynamic = val.lower() in ("1", "yes", "true", "on")
    if dynamic:
        return False
    if not user:
        return True
    return user in ("root", "0")


def find_root_services(deb_path: str) -> list:
    """List bundled systemd system service units that would run as root."""
    try:
        tar_bytes = subprocess.run(["dpkg-deb", "--fsys-tarfile", deb_path],
                                   capture_output=True, check=True).stdout
    except (subprocess.CalledProcessError, FileNotFoundError):
        return []
    import io
    import tarfile
    offenders = []
    with tarfile.open(fileobj=io.BytesIO(tar_bytes)) as tf:
        for m in tf.getmembers():
            if not m.isfile():
                continue
            name = m.name.lstrip("./")
            if not (name.endswith(".service") and name.startswith(_SYSTEMD_SYSTEM_DIRS)):
                continue
            f = tf.extractfile(m)
            if f is None:
                continue
            if service_runs_as_root(f.read().decode("utf-8", "replace")):
                offenders.append(name)
    return offenders


def check_desktop(deb_path: str) -> bool:
    try:
        result = subprocess.run(["dpkg-deb", "-c", deb_path],
                                capture_output=True, text=True, check=True)
        return any(line.endswith(".desktop") for line in result.stdout.splitlines())
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("ERROR: dpkg-deb not found. Is dpkg-deb installed?", file=sys.stderr)
        sys.exit(1)


def extract_metadata(deb_path: str) -> dict:
    fields = ["Package", "Version", "Architecture", "Maintainer"]
    values = {}
    for field in fields:
        try:
            result = subprocess.run(["dpkg-deb", "-f", deb_path, field],
                                    capture_output=True, text=True, check=True)
            values[field] = result.stdout.strip()
        except subprocess.CalledProcessError:
            values[field] = ""

    maintainer = values.get("Maintainer", "")
    email = extract_email(maintainer)

    return {
        "package": values.get("Package", ""),
        "version": values.get("Version", ""),
        "architecture": values.get("Architecture", ""),
        "maintainer": maintainer,
        "maintainer_email": email,
    }


def extract_email(maintainer: str) -> str:
    start = maintainer.find("<")
    end = maintainer.find(">")
    if start != -1 and end != -1:
        return maintainer[start + 1:end]
    return maintainer


def is_valid_package_name(name: str) -> bool:
    if len(name) < 2:
        return False
    valid_chars = set("abcdefghijklmnopqrstuvwxyz0123456789.+-")
    first_chars = set("abcdefghijklmnopqrstuvwxyz0123456789")
    if name[0] not in first_chars:
        return False
    return all(c in valid_chars for c in name)


def branch_name(meta: dict) -> str:
    ts = int(time.time())
    return f"publish/{meta['package']}-{meta['version']}-{ts}"


def check_version_newer(meta: dict):
    packages_url = "https://cardputerzero.github.io/packages/dists/stable/main/binary-arm64/Packages"
    try:
        req = urllib.request.Request(packages_url)
        resp = urllib.request.urlopen(req, timeout=10)
        content = resp.read().decode()
    except Exception:
        return

    in_our_package = False
    existing_version = None

    for line in content.splitlines():
        if line.startswith("Package: "):
            in_our_package = (line[len("Package: "):] == meta["package"])
        if in_our_package and line.startswith("Version: "):
            ver = line[len("Version: "):]
            if existing_version is None or compare_versions(ver, existing_version) > 0:
                existing_version = ver
        if line == "":
            in_our_package = False

    if existing_version:
        if compare_versions(meta["version"], existing_version) <= 0:
            print(f"ERROR: {meta['version']} is not newer than existing version {existing_version}", file=sys.stderr)
            print("  Use `czdev bump` to check the next version.", file=sys.stderr)
            sys.exit(1)
        print(f"  ✓ {meta['version']} is newer than existing {existing_version}")
    else:
        print("  ✓ New package (no existing version found)")
        print("    NOTE: first releases are reviewed by a human. After the PR is")
        print("    created, post a short video of the app running on a real device")
        print("    in the PR comments; a maintainer will merge it after watching.")
        print("    Later version updates merge automatically.")


def compare_versions(a: str, b: str) -> int:
    def parse(v):
        return [int(x) for x in re.split(r'[.\-~]', v) if x.isdigit()]

    pa, pb = parse(a), parse(b)
    if pa > pb:
        return 1
    elif pa < pb:
        return -1
    return 0


def check_git_installed():
    try:
        subprocess.run(["git", "--version"], capture_output=True, check=True)
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("git is not installed.\n  Install: https://git-scm.com/downloads", file=sys.stderr)
        sys.exit(1)


def run_cmd_in(cwd: Path, cmd: list):
    result = subprocess.run(cmd, cwd=str(cwd), capture_output=True, text=True)
    if result.returncode != 0:
        cmd_str = " ".join(cmd)
        print(f"ERROR: {cmd_str} failed:\n{result.stderr}", file=sys.stderr)
        sys.exit(1)
